src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/InlayHints/InlayHintService.cs

```diff
@@ -0,0 +1,126 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT license. See License.txt in the project root for license information.
+
+using System.ComponentModel.Composition;
+using System.Diagnostics;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Razor.Language;
+using Microsoft.AspNetCore.Razor.LanguageServer.Common;
+using Microsoft.AspNetCore.Razor.LanguageServer.Extensions;
+using Microsoft.AspNetCore.Razor.LanguageServer.Protocol;
+using Microsoft.AspNetCore.Razor.LanguageServer.Semantic;
+using Microsoft.AspNetCore.Razor.PooledObjects;
+using Microsoft.CodeAnalysis.Razor.Workspaces.Extensions;
+using Microsoft.VisualStudio.LanguageServer.Protocol;
+using Newtonsoft.Json.Linq;
+
+namespace Microsoft.AspNetCore.Razor.LanguageServer.InlayHints;
+
+[Export(typeof(IInlayHintService))]
+[method: ImportingConstructor]
+internal sealed class InlayHintService(IRazorDocumentMappingService documentMappingService) : IInlayHintService
+{
+    private readonly IRazorDocumentMappingService _documentMappingService = documentMappingService;
+
+    public async Task<InlayHint[]?> GetInlayHintsAsync(IClientConnection clientConnection, VersionedDocumentContext documentContext, Range range, CancellationToken cancellationToken)
+    {
+        var codeDocument = await documentContext.GetCodeDocumentAsync(cancellationToken).ConfigureAwait(false);
+        var csharpDocument = codeDocument.GetCSharpDocument();
+
+        // We are given a range by the client, but our mapping only succeeds if the start and end of the range can both be mapped
+        // to C#. Since that doesn't logically match what we want from inlay hints, we instead get the minimum range of mappable
+        // C# to get hints for. We'll filter that later, to remove the sections that can't be mapped back.
+        if (!_documentMappingService.TryMapToGeneratedDocumentRange(csharpDocument, range, out var projectedRange) &&
+            !RazorSemanticTokensInfoService.TryGetMinimalCSharpRange(codeDocument, range, out projectedRange))
```

Comment:
>RazorSemanticTokensInfoService.TryGetMinimalCSharpRange [](http://example.com/codeflow?start=13&length=55)

For my education - what would be some case(s) when IRazorDocumentMappingService fails to map but RazorSemanticTokensInfoService succeeds?

Reply:
Most of the time the IDE is going to ask us for hints in the range of whatever is visible in the document window, so for a small document that's going to be (0,0) to EOF. If neither (0,0) or EOF are C# code (which I think is technically impossible in Razor) then our mapping with fail, so we need to use the minimal range instead.

In fact, the only time the mapping service is likely to win here is in a very large `@code` block that is longer than the number of lines shown in the editor window, when the user scrolls it in to view.

Reply:
Thanks!